File: Base/sremu 3/server/trunk/src/Auth/LoginSocket.cpp
/* 
 * Copyright (C) 2005-2008 SREmu <http://www.sremu.org/>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <iostream>
#include "LoginSocket.h"

void LoginSocket::ReceiveThread(LPVOID s) 
{
	LoginSocket Client(*((SOCKET*)s));
	Client.Init();
	Client.Receive();
}

void LoginSocket::ProcessData(unsigned char* data, int size) 
{
	Reader.SetBuffer(data);

	int offset = 0;
	while(offset < size) {

		unsigned short psize  = Reader.ReadWord() + 6;
		unsigned short opcode = Reader.ReadWord();
		Reader.Skip(2);

		switch(opcode) 
		{
			case LOGIN_CLIENT_KEEP_ALIVE:
			case LOGIN_CLIENT_ACCEPT_HANDSHAKE:
				break;

			case LOGIN_CLIENT_LAUNCHER:
				SendLauncherInfo();
				break;

			case LOGIN_CLIENT_INFO:
				SendServerInfo();
				break;

			case LOGIN_CLIENT_PATCH_REQUEST:
				SendPatchInfo();
				break;

			case LOGIN_CLIENT_SERVERLIST_REQUEST:
				SendServerList();
				break;

			case LOGIN_CLIENT_AUTH:
				OnLogin();
				break;

			default:
				printf("Unknown opcode: %.4x\n", opcode);
		}
		Reader.Reset();
		Reader.Skip(psize);
		offset += psize;
	}
	Reader.Reset();
}

void LoginSocket::Init() 
{
	Writer.Create(LOGIN_SERVER_HANDSHAKE);
		Writer.WriteByte(1);
	Writer.Finalize();
	Send(Writer.Buffer, Writer.Size());
}

void LoginSocket::SendServerInfo() 
{
	unsigned char name[] = "GatewayServer";
	int namelen = strlen((char*)name);

	Writer.Create(LOGIN_SERVER_INFO);
		Writer.WriteWord(namelen);
		Writer.WriteString(name, namelen);
		Writer.WriteByte(0);
	Writer.Finalize();
	Send(Writer.Buffer, Writer.Size());
}

void LoginSocket::SendLauncherInfo() 
{
	Writer.Create(LOGIN_SERVER_PATCH_INFO);
		Writer.WriteDWord(0x04000101);
		Writer.WriteByte(0xA1);
	Writer.Finalize();
	Send(Writer.Buffer, Writer.Size());

	Writer.Create(LOGIN_SERVER_LAUNCHER);

		mysqlpp::StoreQueryResult res = db.Query("select * from news order by id desc").store();

		Writer.WriteByte(0x00);
		Writer.WriteByte(res.num_rows() < 5 ? res.num_rows() : 4);

		for(unsigned int i = 0; i < res.num_rows(); ++i) 
		{
			unsigned char* head = (unsigned char*)res[i]["head"].c_str();
			unsigned int headlen = strlen((char*)head);

			unsigned char* text = (unsigned char*)res[i]["text"].c_str();
			unsigned int textlen = strlen((char*)text); 

			/* Some data is incomplete here. */
			Writer.WriteWord(strlen(res[i]["head"]));
			Writer.WriteString(head, headlen);
			Writer.WriteWord(strlen(res[i]["text"]));
			Writer.WriteString(text, textlen);
			Writer.WriteWord(0x07D8);
			Writer.WriteByte(atoi(res[i]["month"]));
			Writer.WriteByte(0x00);
			Writer.WriteByte(atoi(res[i]["day"]));
			Writer.WriteByte(0x00);
			Writer.WriteDWord(0x0025000B);
			Writer.WriteDWord(0xAFC00012);
			Writer.WriteWord(0x35D2);
		}

	Writer.Finalize();
	Send(Writer.Buffer, Writer.Size());	
}
void LoginSocket::SendPatchInfo() 
{
	Writer.Create(LOGIN_SERVER_PATCH_INFO);
		Writer.WriteDWord(0x05000101);
		Writer.WriteByte(0x20);
	Writer.Finalize();
	Send(Writer.Buffer, Writer.Size());

	Writer.Create(LOGIN_SERVER_PATCH_INFO);
		Writer.WriteDWord(0x01000100);
		Writer.WriteDWord(0x00050628);
		Writer.WriteWord(0);
		Writer.WriteByte(2);
	Writer.Finalize();
	Send(Writer.Buffer, Writer.Size());

	Writer.Create(LOGIN_SERVER_PATCH_INFO);
		Writer.WriteDWord(0x05000101);
		Writer.WriteByte(0x60);
	Writer.Finalize();
	Send(Writer.Buffer, Writer.Size());

	Writer.Create(LOGIN_SERVER_PATCH_INFO);
		Writer.WriteDWord(0x02000300);
		Writer.WriteWord(0x0200);
	Writer.Finalize();
	Send(Writer.Buffer, Writer.Size());

	Writer.Create(LOGIN_SERVER_PATCH_INFO);
		Writer.WriteDWord(0x00000101);
		Writer.WriteByte(0xA1);
	Writer.Finalize();
	Send(Writer.Buffer, Writer.Size());

	Writer.Create(LOGIN_SERVER_PATCH_INFO);
		Writer.WriteWord(0x0100);
	Writer.Finalize();
	Send(Writer.Buffer, Writer.Size());
}

void LoginSocket::SendServerList() 
{
	unsigned char name[] = "SRO_Global_TestBed";
	int namelen = strlen((char*)name);

	Writer.Create(LOGIN_SERVER_LIST);

		Writer.WriteByte(0x01);
		Writer.WriteByte(0x15);
		
		Writer.WriteWord(namelen);
		Writer.WriteString(name, namelen);
		Writer.WriteByte(0);

		mysqlpp::StoreQueryResult res = db.Query("select * from servers").store();
		for(unsigned int i = 0; i < res.num_rows(); ++i) 
		{
			unsigned char* name = (unsigned char*)res[i]["name"].c_str();
			unsigned int namelen = strlen((char*)name);

			Writer.WriteByte(1);
			Writer.WriteWord(atoi(res[i]["id"]));
			Writer.WriteWord(strlen(res[i]["name"]));
			Writer.WriteString(name, namelen);
			Writer.WriteWord(atoi(res[i]["users_current"]));
			Writer.WriteWord(atoi(res[i]["users_max"]));
			Writer.WriteByte(atoi(res[i]["state"]));

		}
		Writer.WriteByte(0);

	Writer.Finalize();
	Send(Writer.Buffer, Writer.Size());
}

void LoginSocket::OnLogin() 
{
	Reader.Skip(1);

	unsigned short userlen	= Reader.ReadWord();
	unsigned char* user		= new unsigned char[userlen+1];
	memset(user, 0x00, userlen+1);
	Reader.ReadString(user, userlen);

	unsigned short passlen	= Reader.ReadWord();
	unsigned char* pass		= new unsigned char[passlen+1];
	memset(pass, 0x00, passlen+1);
	Reader.ReadString(pass, passlen);

	unsigned short serverid = Reader.ReadWord(); 
	printf("User attempting to log in...\n");

	Writer.Create(LOGIN_SERVER_AUTH_INFO);
	
		mysqlpp::StoreQueryResult res = 
			db.Query("select * from users where username='%s'", (char*)user).store();

		if(res.num_rows()!=0) 
		{
			if(strcmp((char*)pass, res[0]["password"].c_str())==0) 
			{
				db.Query("update users set failed_logins=0 where username='%s'", (char*)user).execute();

				mysqlpp::StoreQueryResult res = 
					db.Query("select * from servers where id=%d", serverid).store();

				Writer.WriteByte(0x01);
				Writer.WriteDWord(0x1234);
				Writer.WriteWord(strlen(res[0]["ip"]));
				Writer.WriteString((unsigned char*)res[0]["ip"].c_str(), strlen(res[0]["ip"]));
				Writer.WriteWord(atoi(res[0]["port"]));	
			} 
			else 
			{
				Writer.WriteByte(0x02);

				if(atoi(res[0]["failed_logins"])==5) 
				{
					Writer.WriteByte(0x06);
				}  
				else 
				{
					Writer.WriteByte(0x01);	
					Writer.WriteDWord(res[0]["failed_logins"]+1);
					Writer.WriteDWord(0x05);

					db.Query("update users set failed_logins=%d", res[0]["failed_logins"]+1).execute();
				}
			}
		} 
		else 
		{
			Writer.WriteByte(0x02);
			Writer.WriteByte(0x04);

		}

	Writer.Finalize();
	Send(Writer.Buffer, Writer.Size());

	delete[] user;
	delete[] pass;
}

File: Base/Old/Source/LoginServer/Login/LoginSocket.cs
/*    <DarkEmu LoginServer>
    Copyright (C) <2011>  <DarkEmu>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

using System;
using System.Collections.Generic;
using System.Text;

namespace DarkEmu_LoginServer
{
    class LoginSocket
    {
        public static bool debug = false;

        unsafe public static void ProcessData(byte[] buffer, int ClientIndex)
        {
            TPacket* tmpPacket = Silkroad.ToTPacket(buffer);
            if (debug)            
                Console.WriteLine("[ProcessData][{0:X}][{1} bytes][Index {2}]\n{3}\n", tmpPacket->opcode, tmpPacket->size,ClientIndex,BitConverter.ToString(buffer,6,tmpPacket->size).Replace('-',' '));
           
            switch (tmpPacket->opcode)
            {
                case CLIENT_OPCODES.LOGIN_CLIENT_KEEP_ALIVE:
                case CLIENT_OPCODES.LOGIN_CLIENT_ACCEPT_HANDSHAKE:
                    break;
                case CLIENT_OPCODES.LOGIN_CLIENT_INFO:
                    SendServerInfo(ClientIndex);
                    break;
                case CLIENT_OPCODES.LOGIN_CLIENT_PATCH_REQUEST:
                    SendPatchInfo(ClientIndex);
                    break;
                case CLIENT_OPCODES.LOGIN_CLIENT_LAUNCHER:
                    SendLauncherInfo(ClientIndex);
                    break;
                case CLIENT_OPCODES.LOGIN_CLIENT_SERVERLIST_REQUEST:
                    SendServerList(ClientIndex);
                    break;
                case CLIENT_OPCODES.LOGIN_CLIENT_AUTH:
                    /*if (debug)
                        Debugx.DumpBuffer(buffer, 1, tmpPacket->opcode, tmpPacket->size);*/
                    SendLogin(buffer, ClientIndex);
                    break;
                case CLIENT_OPCODES.LOGIN_CLIENT_LAUNCHER_UNK1:
                    SendLauncherUnk1(ClientIndex);
                    break;
                case CLIENT_OPCODES.LOGIN_CLIENT_AUTH_UNK1:
                    SendServerUnk1(ClientIndex);
                    break;
                default:
                    //Debugx.DumpBuffer(buffer, 1, tmpPacket->opcode, tmpPacket->size);
                    Console.WriteLine("default Opcode: {0}", tmpPacket->opcode);
                    break;
            }

        }

        private static void SendServerInfo(int ClientIndex)
        {
            string name = "GatewayServer";
            ushort namelen = (ushort)name.Length;

            PacketWriter writer = new PacketWriter();

            writer.SetOpcode(SERVER_OPCODES.LOGIN_SERVER_INFO);
            writer.AppendWord(namelen);
            writer.AppendString(false, name);
            writer.AppendByte(0x00);
            ServerSocket.Send(writer.getWorkspace(), ClientIndex);
        }

        private static void SendPatchInfo(int ClientIndex)
        {
            PacketWriter writer = new PacketWriter();

            writer.SetOpcode(SERVER_OPCODES.LOGIN_SERVER_PATCH_INFO);
            writer.AppendDword(0x05000101);
            writer.AppendByte(0x20);
            ServerSocket.Send(writer.getWorkspace(), ClientIndex);

            writer.SetOpcode(SERVER_OPCODES.LOGIN_SERVER_PATCH_INFO);
            writer.AppendDword(0x01000100);
            writer.AppendDword(0x00050628);
            writer.AppendWord(0x00);
            writer.AppendByte(0x02);
            ServerSocket.Send(writer.getWorkspace(), ClientIndex);

            writer.SetOpcode(SERVER_OPCODES.LOGIN_SERVER_PATCH_INFO);
            writer.AppendDword(0x05000101);
            writer.AppendByte(0x60);
            ServerSocket.Send(writer.getWorkspace(), ClientIndex);

            writer.SetOpcode(SERVER_OPCODES.LOGIN_SERVER_PATCH_INFO);
            writer.AppendDword(0x02000300);
            writer.AppendWord(0x0200);
            ServerSocket.Send(writer.getWorkspace(), ClientIndex);

            writer.SetOpcode(SERVER_OPCODES.LOGIN_SERVER_PATCH_INFO);
            writer.AppendDword(0x00000101);
            writer.AppendByte(0xA1);
            ServerSocket.Send(writer.getWorkspace(), ClientIndex);

            writer.SetOpcode(SERVER_OPCODES.LOGIN_SERVER_PATCH_INFO);
            writer.AppendWord(0x0100);
            ServerSocket.Send(writer.getWorkspace(), ClientIndex);
        }

        private static void SendLauncherInfo(int ClientIndex)
        {
            PacketWriter writer = new PacketWriter();

            writer.SetOpcode(SERVER_OPCODES.LOGIN_SERVER_PATCH_INFO);
            writer.AppendDword(0x04000101);
            writer.AppendByte(0xA1);
            ServerSocket.Send(writer.getWorkspace(), ClientIndex);

            writer.SetOpcode(SERVER_OPCODES.LOGIN_SERVER_LAUNCHER);
            writer.AppendByte(0x00);

            writer.AppendByte(DatabaseCore.News.NumberOfNews);
            for (int i = 0; i < DatabaseCore.News.NumberOfNews; i++)
            {

                writer.AppendWord((ushort)DatabaseCore.News.Head[i].Length);
                writer.AppendString(false,DatabaseCore.News.Head[i]);

                writer.AppendWord((ushort)DatabaseCore.News.Text[i].Length);
                writer.AppendString(false, DatabaseCore.News.Text[i]);

                writer.AppendWord(0x07D8);

                writer.AppendByte(DatabaseCore.News.Month[i]);
                writer.AppendByte(0x00);

                writer.AppendByte(DatabaseCore.News.Day[i]);
                writer.AppendByte(0x00);

                writer.AppendDword(0x0025000B);
                writer.AppendDword(0xAFC00012);
                writer.AppendWord(0x35D2);

            }
            ServerSocket.Send(writer.getWorkspace(), ClientIndex);
        }
        private static void SendLauncherUnk1(int ClientIndex)
        {
            PacketWriter2 writer = new PacketWriter2(SERVER_OPCODES.LOGIN_SERVER_LAUNCHER_UNK1, 1026, true);
            //01 01 04 02
            writer.AddByte(0x01);
            writer.AddByte(0x01);
            writer.AddByte(0x04);
            writer.AddByte(0x02);
            ServerSocket.Send(writer.GetBytes(), ClientIndex);
        }
        private static void SendServerUnk1(int ClientIndex)
        {
            /*
             43 00 L
             07 A1 Opcode
             00 00 Secu
             
             03 00 count
             
             11 00 
             67 73 70 6B 72 31 2E 6A 6F 79 6D 61 78 2E 63 6F 6D BD string_ip
             32 01 port
             */
            PacketWriter2 writer = new PacketWriter2(SERVER_OPCODES.LOGIN_SERVER_AUTH_UNK1);
            writer.AddByte(0x01);
            writer.AddByte(0x00);
            string str_ip = "127.0.0.1";
            writer.AddByte(0x09);
            writer.AddByte(0x00);
            writer.AddString(str_ip, str_ip.Length);
            writer.AddByte(0x3D);
            writer.AddByte(0xA4);
            ServerSocket.Send(writer.GetBytes(), ClientIndex);

        }
        private static void SendServerList(int ClientIndex)
        {
            PacketWriter writer = new PacketWriter();

            string name = "SRO_Global_TestBed";
            ushort namelen = (ushort)name.Length;

            writer.SetOpcode(SERVER_OPCODES.LOGIN_SERVER_LIST);

            writer.AppendByte(0x01);
            writer.AppendByte(0x15);

            writer.AppendWord(namelen);
            writer.AppendString(false, name);
            writer.AppendByte(0x00);

            for (int i = 1; i < DatabaseCore.Server.NumberOfServer; i++)
            {
                writer.AppendByte(0x01);
                writer.AppendWord(DatabaseCore.Server.ServerId[i]);
                writer.AppendWord((ushort)(DatabaseCore.Server.ServerName[i].Length+1));
                writer.AppendByte(0x31); // flag xD usa
                writer.AppendString(false, DatabaseCore.Server.ServerName[i]);
                writer.AppendWord(DatabaseCore.Server.CurUser[i]);
                writer.AppendWord(DatabaseCore.Server.MaxUser[i]);
                writer.AppendByte(DatabaseCore.Server.ServerState[i]);
            }
            writer.AppendByte(0x00);
            ServerSocket.Send(writer.getWorkspace(), ClientIndex);
        }        

        private static void SendLogin(byte[] buffer, int ClientIndex)
        {
            /*
            16 00 L
            02 61 Opcode
            00 00 Secu
            
            12 Leng_total
            07 00 Leng_String
            63 61 72 6C 6F 73 78 = carlosx
            07 00 Leng_String
            63 61 72 6C 6F 73 78 = carlosx
            
            FF unk
            03 ServerID
            */
            PacketWriter writer = new PacketWriter();
            PacketReader reader = new PacketReader(buffer, buffer.Length);
            reader.ModifyIndex(7);

            ushort userlen = reader.ReadWord();
            string user = reader.ReadString(false, userlen);

            ushort passlen = reader.ReadWord();
            string pass = reader.ReadString(false, passlen);

            reader.ReadByte();
            ushort serverid = reader.ReadByte();

            Console.WriteLine("User:{0} - Password:{1} - ServerId:{2}",user,pass,serverid);

            int UserIndex = DatabaseCore.User.GetIndexByName(user);

            if (DatabaseCore.User.Banned[UserIndex])
            {
                ServerSocket.DisconnectSocket(ClientIndex);
                return;
            }
            else if (DatabaseCore.User.Online[UserIndex])
            {
                writer.SetOpcode(SERVER_OPCODES.LOGIN_SERVER_AUTH_INFO);
                writer.AppendWord(0x302);
                ServerSocket.Send(writer.getWorkspace(), ClientIndex);
                ServerSocket.DisconnectSocket(ClientIndex);
            }
            else if (DatabaseCore.User.User[UserIndex] != user || DatabaseCore.User.Password[UserIndex] != pass)
            {
                writer.SetOpcode(SERVER_OPCODES.LOGIN_SERVER_AUTH_INFO);

                DatabaseCore.User.FailedLogins[UserIndex]++;
                DatabaseCore.WriteQuery("UPDATE user SET failed_logins = '{0}' WHERE id = '{1}'", DatabaseCore.User.FailedLogins[UserIndex], DatabaseCore.User.User[UserIndex]); 

                writer.AppendWord(0x0102);
                writer.AppendByte(DatabaseCore.User.FailedLogins[UserIndex]);
                writer.AppendDword(0x03000000);
                writer.AppendWord(0x00);
                writer.AppendByte(0x00);

                ServerSocket.Send(writer.getWorkspace(), ClientIndex);

                if (DatabaseCore.User.FailedLogins[UserIndex] == 3)
                {
                    DatabaseCore.User.FailedLogins[UserIndex] = 0;
                    DatabaseCore.WriteQuery("UPDATE user SET failed_logins = '{0}' WHERE id = '{1}'", DatabaseCore.User.FailedLogins[UserIndex], DatabaseCore.User.User[UserIndex]); 

                    ServerSocket.DisconnectSocket(ClientIndex);
                }
            }
            else
            {
                int ServerIndex = DatabaseCore.Server.GetIndexById(serverid);

                if (DatabaseCore.Server.MaxUser[ServerIndex] <= DatabaseCore.Server.CurUser[ServerIndex])
                {
                    writer.SetOpcode(0xA103);
                    writer.AppendWord(0x402);
                    ServerSocket.Send(writer.getWorkspace(), ClientIndex);
                    return;
                }
                else
                {
                    writer.SetOpcode(SERVER_OPCODES.LOGIN_SERVER_AUTH_INFO);
                    writer.AppendByte(0x01);
                    writer.AppendDword(0x1234);

                    ushort leng_ip = (ushort)DatabaseCore.Server.ServerIp[ServerIndex].Length;
                    string str_ip = DatabaseCore.Server.ServerIp[ServerIndex];
                    ushort port = DatabaseCore.Server.ServerPort[ServerIndex];

                    writer.AppendWord(leng_ip);
                    writer.AppendString(false, str_ip);
                    writer.AppendWord(port);
                    ServerSocket.Send(writer.getWorkspace(), ClientIndex);
                    Console.WriteLine("Lengip:{0} - Ip:{1} - Port:{2}",leng_ip,str_ip,port);

                    //DatabaseCore.Server.CurUser[ServerIndex]++;
                    //DatabaseCore.WriteQuery("UPDATE server SET users_current = '{0}' WHERE name = '{1}'", DatabaseCore.Server.CurUser[ServerIndex], DatabaseCore.Server.ServerName[ServerIndex]);
                }
            }
        }
    }
}
